### PR TITLE
the cell is landmark for a form

### PR DESCRIPTION
**Pull request by tonyfast**
this pull request began as an exploration of potential configurations for an accessible cell form. 

* the cell is significant enough feature to be a landmark.
* the cell has a `form`
* the cell has input and output regions
* the cell complies with web accessibility guidelines

the implementation uses semantic elements. Another implementation could use the implicit roles and proper ARIA to create an explicit accessible experience. features will need to be added and they should include accessible elements that are prescribed by references like [ARIA Authoring Practices Guide](https://www.w3.org/WAI/ARIA/apg/patterns/)

example outputs below:

* https://iota-school.github.io/notebooks-for-all/branch/extent/exports/html/stsci_example_notebook-smol.html
* https://iota-school.github.io/notebooks-for-all/branch/extent/exports/html/Imaging_Sky_Background_Estimation-smol.html
* https://iota-school.github.io/notebooks-for-all/branch/extent/exports/html/lorenz-executed-smol.html

this work includes testing on NVDA, VoiceOver, Orca, and ChromeVox.

## the cell landmark

the cell is a landmark with `role=region` when there is an ARIA tag. an implementation would choose whether some, all, or none of the cells are labeled.

## the cell form

the cell `form` is an empty element. all the content exists outside the `form` to allow other `form`s; `form` elements cannot be nested. to tether a cell's form associated elements to. the `document.forms` object provides nominal and ordinal access to each form.

## the cell input region

the cell input `fieldset` has an implicit `role=group`. the `disabled` attribute applies to all form descendants. a static, resting cell then is `<fieldset disabled>` which allows for custom styling based on state.

## the cell output region

the cell output holds computed values from the `input`. we represent these with the `output` tag which implicitly is an aria live element. instead we apply `role=presentation` to reduce the screen verbosity.

## the id patterns

a notebook is data and a document. we use a consistent json pointer naming convention that would provide consistent access to either the json document or html document.

## styling

styling is a non-goal of this pull request it uses minimum styling to convey the message of the notebook. we do demonstrate how semantic css selectors naturally facilitate formatting the document. using flex box will help with reactive experiences at high zoom magnifications.

## references

* https://www.w3.org/WAI/ARIA/apg/patterns/landmarks/examples/general-principles.html
  * use `aside` for footnotes

### `output`

* https://www.scottohara.me/blog/2019/07/10/the-output-element.html
* https://developer.mozilla.org/en-US/docs/Web/HTML/Element/section

### `fieldset` 

* https://adrianroselli.com/2022/07/use-legend-and-fieldset.html
* https://www.tpgi.com/fieldsets-legends-and-screen-readers-again/

### CODE REVIEW

**Comment by tonyfast** [comment]
it seems the `<section aria-labelledby=X><fieldset><legend id=X>` configuration will always repeat the legend contents on VoiceOver. this would only be favorable in an editing mode.

the `fieldset` was added because we needed to group multiple projections of the same input. right now, the markdown cells don't use the output region because the markdown source and output are in the input region. if the rendering is moved to the output region then the experience would be good. in that sense, the markdown to html operation is the kernel, it could be remote or local.

basically, we need an input group for any real world applications. the input group could be a fieldset, tabs, accordion, or other accessible patterns.

**Comment by tonyfast** [comment]
this working is tedious, but the questions are changing and this is good. the execution count is still a stinking nuisance .

## the execution count

the cell is form and effectively the execution count, cell index, or cell (insert custom name) is the label for it. the label needs to be exposed, but not too much.

does the execution count landmark the cell, cell input, or cell output?

the label is actually a generated output. `<output name="execution_count"><label>29</label></output>`

## the rendered markdown and the rendered code

i had markdown output being rendering in the input group. @bollwyvl said that the rendered markdown is an output, and i agree with that. now i don't know if the rendered highlight pygments code is part of the input or the output. 

> as i :duck: this, i think pygments representation is in the input group because it is a dual of the input text. the markdown view is the source interpretted, but pygments does no interpretation so there is no information loss.

**Comment by tonyfast** [comment]
the redundancy of In[x] and Out[x] only provides noise on the screen reader. there is no reason those numbers will ever differ and one of the them is superfluous while the other is a `label`. the only way to advance the execution count is to interact with a kernel that increments it. in theory, the Out[x] is sufficient enough to label an executed cell.

**Comment by tonyfast** [comment]
i just added the smol variant that is meant to be accessible only and not using jupyter conventions, no pygments, just bare tags. this variant lets us understand how create the notebook variant. in this variant, all cells are labelled ordinally and the execution count is applied to only the output. its clear thought that if there is only one cell then there is a consistent naming for any genus of cell. the cell index in a notebook is a default, but we could custom name cells like observable offers.

**Comment by tonyfast** [comment]
for this idea we need a canonical toggle for read mode to execute mode. if the starting point is accessible then the application can be developed accessibly, and accessibility guides design decisions.

in smol, i ripped out the input `fieldset` because it was causing some verbosity issues on the screen readers. it turns out the `fieldset[name=input]` was not the problem. when we were using it we were giving markdown and code cells different naming heuristics. duh, there was a heuristic in the way! we gotta get rid of everything stinking heuristic.

`fieldset[name=input]` comes back so that `<fieldset name=input disabled>` can disable the form and `fieldset[name=input]:disabled` is the toggle for our primary states. we are doing this in the `sections` template already, but the labeling is bunk.

we will now label each section by a cell name with the an ordinal default that could be modified later. each input and output should be marked.

**Comment by tonyfast** [comment]
When the cell label is hidden so is the prefix to the screen reader. The cell, in, out prefixes need to be in the dom.

**Comment by smythp** [comment]
Sure, I'll take a look now.

**Comment by tonyfast** [comment]
should have mentioned to use this link https://iota-school.github.io/notebooks-for-all/branch/extent/exports/html/lorenz-executed-smol.html

**Comment by tonyfast** [comment]
no, these are the right one

> 
>  * https://iota-school.github.io/notebooks-for-all/branch/extent/exports/html/stsci_example_notebook-smol.html
> 
>  * https://iota-school.github.io/notebooks-for-all/branch/extent/exports/html/Imaging_Sky_Background_Estimation-smol.html
> 
>  * https://iota-school.github.io/notebooks-for-all/branch/extent/exports/html/lorenz-executed-smol.html

File: nbconvert_html5/templates/semantic-forms/smol.html.j2 [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+</select>
+{%- endmacro -%}
+
+
+{%- macro cell_output(cell, nb, parts) -%}
+{%- set count = nb["cells"].index(cell)-%}{%- set ID = "/cells/" + str(count) -%}
+{% set CODE = cell.cell_type == "code" %}
+
+<fieldset name="outputs" id="{{ID}}/outputs" {% if CODE and not cell.outputs %}hidden{% endif %} {% if not CODE %}role="presentation"{% endif %}>
+    <legend id={{ID}}/execution_count {% if not CODE %}hidden{% endif %}>Out&nbsp;<output name="execution_count"  role="presentation">{{cell.execution_count or ""}}</output></legend>
```
Comment (tonyfast) [inline nbconvert_html5/templates/semantic-forms/smol.html.j2:126]: NVDA repeats the legend text after announcing the group. we need to use the same label trick used on the input.

Comment (tonyfast) [reply nbconvert_html5/templates/semantic-forms/smol.html.j2]: the label think worked to fix the NVDA redundancy. Out is accidentally visually there and i just pushed a change to fix that.

File: nbconvert_html5/templates/semantic-forms/smol.html.j2 [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+    }
+</style>{% endblock css %}
+{% block any_cell scoped %}
+{%- set count = nb["cells"].index(cell)-%}{%- set ID = "/cells/" + str(count) -%}
+{%- set CODE = cell.cell_type == "code" -%}
+{%- set tags = celltags(cell) -%}
+<form hidden id="{{ID}}" name="{{ID}}" method="POST" action="{{cell.cell_type}}" aria-labelledby={{ID}}/cell_count>
+    <label id={{ID}}/cell_count>Cell&nbsp;<output name="cell_count" role="presentation">{{count + 1}}</output></label>
+</form>
+<section aria-labelledby="{{ID}}/cell_count" class="cell {{cell.cell_type}}{% if tags %} {{tags}}{% endif %}" id="{{count}}">
```
Comment (tonyfast) [inline nbconvert_html5/templates/semantic-forms/smol.html.j2:86]: i'm not happy. i can't get VoiceOver reliably to announce cells. they are only announced when there are at least to block nodes to be announced.

bundling contiguous markdown content into a contiguous cell will more reliably get VoiceOver to announce leaving the region.

**Comment by tonyfast** [comment]
hopefully, testing is going okay 😟 . i've done some final QA on VoiceOver, ChromeVox (on Chromebook), and NVDA.

* NVDA sounds pretty good to my untrained ear. love to hear your thoughts.
  * using <kbd>F</kbd> <kbd>Shift + F</kbd> to navigate feels great i think!
* VoiceOver is [not announcing the cells the way i'd expect](https://github.com/Iota-School/notebooks-for-all/pull/63/files#r1218486567). nevertheless, there is little needless repetition. if anything it is too terse.
* ChromeVox does this thing where `readonly` and `disabled` textarea's are repeated and it stinks. so we'd have to have a specific ChromeVox fix i think which doesnt seem worth it.

**Comment by smythp** [comment]
Hi Tony, sorry I got pulled away yesterday afternoon after going through the first notebook. Let me quickly give you notes from the first notebook and I'll finish the rest after meetings today.

---

Honestly, this looks great, at least with Orca, which is what I've tested it with so far.

Some things that are working:

- Really like the landmark movement. Bringing up the list of landmarks and all the cells being there is a joy.
- In Orca, the way the cell is read partially when you hit a landmark is quite nice. I'm pretty sure NVDA won't do this but looking forward to trying it out.
- On a lot of sites Orca gets badly tripped up or starts using a lot of memory. (At least on my machine, using Cinnamon, MATE seems to work better.) Not seeming to get that here.


Some notes:

- On the small example notebook (the first link you shared), there's a cell toward the end that's labeled cell 7. It seems to just be a separator. It's where cell 20 should be. Haven't dug into the source to see why it's occuring.

**Comment by smythp** [comment]
Yeah, I was thinking about that. I think before I would have said yes, we should have it announce the cell type, now I'm slightly less sure, just because I was getting more context than I was expecting while navigating by landmark, and it was evident pretty fast what kind of cell it was.  Still leaning toward having it say the cell type, though, if I had to come down on it.

**Comment by tonyfast** [comment]
it probably makes sense to make this configurable. in our use case, with just python and markdown, we don't need a differentiation. if there were more languages i think we'd want to make it configurable. i opened #64 to talk about configurations if you have ideas.

it is recommended that the aria label is short around 3 words. now it is "Cell 1". saying "Python Cell 1", "SQL Cell 2", "markdown Cell 3" might make more sense in that scenario.

**Comment by smythp** [comment]
I think "code cell" is probably fine if providing the language will be an issue.  It's not like the language is labeled every time for sighted folks (unless I'm misremembering), there's just visual context on whether the cell is a code cell (presence of the "in [0]" text to the left of the cell.) As I recall, the UX also allows "code cell," "markdown cell," and "raw," unless things have changed, so it would also be in line with that.

**Comment by tonyfast** [comment]
> I think "code cell" is probably fine if providing the language will be an issue. It's not like the language is labeled every time for sighted folks (unless I'm misremembering), there's just visual context on whether the cell is a code cell (presence of the "in [0]" text to the left of the cell.) As I recall, the UX also allows "code cell," "markdown cell," and "raw," unless things have changed, so it would also be in line with that.

there is a visual UX with code highlighting. SQL and Python could show visually different. this is a hypothetical notebook we are taling about.  so maybe just the code cells would need to be distinguished.

**Comment by tonyfast** [comment]
i removed the visible cells labels because they are non-canonical. however, without them it appears much harder to talk about our test cases. cell numbers belong somewhere and we'll want to figure that out. with the screen reader on, the lack of cell numbers for markdown becomes really apparent.

i was thinking about the empty cells, or placeholder cells. they probably should be mentioned as empty; maybe `role=separator` has meaning here.  Removing cells from the screen reader would violate an expectation of countability. in theory, we shouldn't ever hide anything without announcing the lack. it leads me to think that a best practice for cells would be always having an output. even if there is stateful work done. an output could say "the foo object has changed state".

**Comment by smythp** [comment]
Yeah, interesting about the empty or seperator cells. I was wondering if I was missing content when going through nonvisually. I suppose they are cells like any other.

+1 to the adding commits, I was just wondering if I was missing something, but that makes sense.

**Comment by tonyfast** [comment]
i think im going to merge this in before i make anymore of a mess. things that need to be done to finish.

* include cell and notebook metadata
* include notebook toolbar
* include visual cell label

**Comment by tonyfast** [comment]
on orca and voiceover the cell numbers are only announced when there is more than one element in a cell. this means a code cells has visible outputs or markdown is non-empty. in fact, we decided to move the markdown outputs outside the output region so that we would announce markdown cells on VoiceOver as reference for position.

one of the things we learn from this experience is that code cells should have outputs. a cell without outputs is not announced in most screen readers. in the case that cells modify state or set variables. it would make sense for that to be announced visually.

i think it is going to be a problem for assistive tech users to relate an output with the variable it is associated with. currently that relationship is visually implicit and should be made semantically explicit.

**Comment by tonyfast** [comment]
closing. please feel free to keep commenting here or open up another issue.
